// @alicloud/mpserverless-sdk/dist/esm/codec.js
import * as tslib_1 from "tslib";
import { HTTPRequestEncoder, HTTPResponseDecoder, PREFIX, BasementClientError, ErrorName, ErrorType, } from '@ant-basement/core';
import crypto from 'crypto-js';
import { isNumber } from 'util';
var MiniProgramHTTPRequestEncoder = (function (_super) {
    tslib_1.__extends(MiniProgramHTTPRequestEncoder, _super);
    function MiniProgramHTTPRequestEncoder(endpoint, spaceId) {
        var _this = _super.call(this, endpoint) || this;
        _this.spaceId = spaceId;
        _this.prefix = PREFIX.CLIENT;
        _this.serviceHeaders = {};
        _this.setBodyField({
            spaceId: spaceId,
        });
        return _this;
    }
    MiniProgramHTTPRequestEncoder.prototype.sign = function (clientSecret) {
        var _a = this.body, spaceId = _a.spaceId, method = _a.method, params = _a.params, token = _a.token, userId = _a.userId;
        var timestamp = Date.now();
        this.setBodyField({
            timestamp: timestamp,
        });
        var signString = '';
        var signObject = {
            spaceId: spaceId,
            timestamp: timestamp,
            method: method,
            params: JSON.stringify(params),
            token: token,
            userId: userId,
        };
        Object.keys(signObject).sort().forEach(function (key) {
            if (signObject[key]) {
                signString = signString + "&" + key + "=" + signObject[key];
            }
        });
        signString = signString.slice(1);
        var sign = crypto.HmacMD5(signString, clientSecret).toString(crypto.enc.Hex);
        this.setServerlessHeaders({ sign: sign });
    };
    MiniProgramHTTPRequestEncoder.prototype.encodeAsHTTPRequestObject = function (additionalObject) {
        if (this.body.params) {
            this.body.params = JSON.stringify(this.body.params);
        }
        return tslib_1.__assign({ url: this.url, data: this.body, method: this.method, headers: this.headers, dataType: 'json' }, additionalObject);
    };
    MiniProgramHTTPRequestEncoder.prototype.clone = function () {
        var encoder = new MiniProgramHTTPRequestEncoder(this.endpoint, this.spaceId);
        encoder.setBodyField(this.body);
        encoder.setBaseHeaders(this.baseHeaders);
        encoder.setServerlessHeaders(this.serverlessHeaders);
        return encoder;
    };
    return MiniProgramHTTPRequestEncoder;
}(HTTPRequestEncoder));
export { MiniProgramHTTPRequestEncoder };
var MiniProgramHTTPResponseDecoder = (function (_super) {
    tslib_1.__extends(MiniProgramHTTPResponseDecoder, _super);
    function MiniProgramHTTPResponseDecoder() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.ERROR_CODES = [11, 12, 13, 14, 19, 20];
        return _this;
    }
    MiniProgramHTTPResponseDecoder.prototype.setStatusAndBody = function (status, body) {
        _super.prototype.setStatusAndBody.call(this, status, body);
        if (this.ERROR_CODES.includes(status)) {
            this._error = BasementClientError.from({
                name: ErrorName.IDE_ERROR,
                code: status.toString(),
                type: ErrorType.IDE_ERROR,
                message: 'request error from Alipay IDE',
            });
        }
    };
    MiniProgramHTTPResponseDecoder.prototype.decode = function (res) {
        this.setHeaders(res.headers || {});
        var body = res.data || res.body;
        if (typeof body === 'string') {
            body = JSON.parse(body);
        }
        if (body && body.data) {
            if (isNumber(body.data.affectedDocs)) {
                body = Object.assign({}, body, tslib_1.__assign({}, body.data));
            }
            else if (Object.prototype.toString.call(body.data) === '[object Object]') {
                body.result = Object.assign({}, body.data);
            }
            else if (Object.prototype.toString.call(body.data) === '[object Array]') {
                body.result = body.data.slice(0);
            }
            else {
                body.result = body.data;
            }
            delete body.data;
        }
        var responseErrorCode = parseInt(res.error, 10);
        if (responseErrorCode) {
            this.setStatusAndBody(responseErrorCode, body);
            return _super.prototype.decode.call(this);
        }
        var responseErrorMessage = res.err;
        if (responseErrorMessage) {
            this.setErrorMessage(responseErrorMessage);
            return _super.prototype.decode.call(this);
        }
        if (res instanceof Error) {
            this.setErrorObject(res);
            return _super.prototype.decode.call(this);
        }
        if (body && typeof body.error === 'object') {
            this.setErrorObject(body.error);
            return _super.prototype.decode.call(this);
        }
        var responseStatusCode = parseInt(res.status || res.statusCode, 10);
        if (responseStatusCode) {
            this.setStatusAndBody(responseStatusCode, body);
            return _super.prototype.decode.call(this);
        }
        this.setStatusAndBody(200, res);
        return _super.prototype.decode.call(this);
    };
    return MiniProgramHTTPResponseDecoder;
}(HTTPResponseDecoder));
export { MiniProgramHTTPResponseDecoder };
