// @alicloud/mpserverless-sdk/dist/umd/codec.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "@ant-basement/core", "crypto-js", "util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    const tslib_1 = require("tslib");
    const core_1 = require("@ant-basement/core");
    const crypto_js_1 = tslib_1.__importDefault(require("crypto-js"));
    const util_1 = require("util");
    class MiniProgramHTTPRequestEncoder extends core_1.HTTPRequestEncoder {
        constructor(endpoint, spaceId) {
            super(endpoint);
            this.spaceId = spaceId;
            this.prefix = core_1.PREFIX.CLIENT;
            this.serviceHeaders = {};
            this.setBodyField({
                spaceId,
            });
        }
        sign(clientSecret) {
            const { spaceId, method, params, token, userId } = this.body;
            const timestamp = Date.now();
            this.setBodyField({
                timestamp,
            });
            let signString = '';
            const signObject = {
                spaceId,
                timestamp,
                method,
                params: JSON.stringify(params),
                token,
                userId,
            };
            Object.keys(signObject).sort().forEach(key => {
                if (signObject[key]) {
                    signString = `${signString}&${key}=${signObject[key]}`;
                }
            });
            signString = signString.slice(1);
            const sign = crypto_js_1.default.HmacMD5(signString, clientSecret).toString(crypto_js_1.default.enc.Hex);
            this.setServerlessHeaders({ sign });
        }
        encodeAsHTTPRequestObject(additionalObject) {
            if (this.body.params) {
                this.body.params = JSON.stringify(this.body.params);
            }
            return {
                url: this.url,
                data: this.body,
                method: this.method,
                headers: this.headers,
                dataType: 'json',
                ...additionalObject,
            };
        }
        clone() {
            const encoder = new MiniProgramHTTPRequestEncoder(this.endpoint, this.spaceId);
            encoder.setBodyField(this.body);
            encoder.setBaseHeaders(this.baseHeaders);
            encoder.setServerlessHeaders(this.serverlessHeaders);
            return encoder;
        }
    }
    exports.MiniProgramHTTPRequestEncoder = MiniProgramHTTPRequestEncoder;
    class MiniProgramHTTPResponseDecoder extends core_1.HTTPResponseDecoder {
        constructor() {
            super(...arguments);
            this.ERROR_CODES = [11, 12, 13, 14, 19, 20];
        }
        setStatusAndBody(status, body) {
            super.setStatusAndBody(status, body);
            if (this.ERROR_CODES.includes(status)) {
                this._error = core_1.BasementClientError.from({
                    name: core_1.ErrorName.IDE_ERROR,
                    code: status.toString(),
                    type: core_1.ErrorType.IDE_ERROR,
                    message: 'request error from Alipay IDE',
                });
            }
        }
        decode(res) {
            this.setHeaders(res.headers || {});
            let body = res.data || res.body;
            if (typeof body === 'string') {
                body = JSON.parse(body);
            }
            if (body && body.data) {
                if (util_1.isNumber(body.data.affectedDocs)) {
                    body = Object.assign({}, body, {
                        ...body.data,
                    });
                }
                else if (Object.prototype.toString.call(body.data) === '[object Object]') {
                    body.result = Object.assign({}, body.data);
                }
                else if (Object.prototype.toString.call(body.data) === '[object Array]') {
                    [...body.result] = body.data;
                }
                else {
                    body.result = body.data;
                }
                delete body.data;
            }
            const responseErrorCode = parseInt(res.error, 10);
            if (responseErrorCode) {
                this.setStatusAndBody(responseErrorCode, body);
                return super.decode();
            }
            const responseErrorMessage = res.err;
            if (responseErrorMessage) {
                this.setErrorMessage(responseErrorMessage);
                return super.decode();
            }
            if (res instanceof Error) {
                this.setErrorObject(res);
                return super.decode();
            }
            if (body && typeof body.error === 'object') {
                this.setErrorObject(body.error);
                return super.decode();
            }
            const responseStatusCode = parseInt(res.status || res.statusCode, 10);
            if (responseStatusCode) {
                this.setStatusAndBody(responseStatusCode, body);
                return super.decode();
            }
            this.setStatusAndBody(200, res);
            return super.decode();
        }
    }
    exports.MiniProgramHTTPResponseDecoder = MiniProgramHTTPResponseDecoder;
});
